File: paleo_workbench/ui/components/badges.py
"""徽章与行内状态（V5-U2）。

PwbBadge：白字深底徽章（BADGE_* / ERROR_RED，三主题 ≥ 4.5:1），
取代此前 5 处各自实现的 status pill。
PwbInlineStatus：小图标 + 文字的行内状态行，tone 驱动图标与文字色，
取代「裸 QLabel 状态文本」方言（降级态渲染统一入口，D6）。
"""
from __future__ import annotations

from PySide6.QtCore import Qt
from PySide6.QtWidgets import QHBoxLayout, QLabel, QWidget

from paleo_workbench import tokens
from paleo_workbench.ui.theme import theme_manager
from paleo_workbench.ui.workstation.common import workstation_icon

BADGE_TONES = ("neutral", "primary", "success", "warning", "error")

#: tone → (默认图标, 图标染色 token 名)
_TONE_ICON: dict[str, tuple[str, str]] = {
    "neutral": ("", "TEXT_SECONDARY"),
    "primary": ("", "PRIMARY"),
    "success": ("circle-check.svg", "SUCCESS"),
    "warning": ("alert-triangle.svg", "WARNING"),
    "process": ("refresh-cw.svg", "ACCENT"),
    "error": ("alert-triangle.svg", "ERROR"),
}


def _repolish(widget: QWidget) -> None:
    style = widget.style()
    style.unpolish(widget)
    style.polish(widget)


class PwbBadge(QLabel):
    """语义色调徽章。tone ∈ {neutral, primary, success, warning, error}。"""

    def __init__(
        self,
        text: str = "",
        *,
        tone: str = "neutral",
        parent: QWidget | None = None,
    ) -> None:
        super().__init__(text, parent)
        self.setObjectName("PwbBadge")
        self.setAlignment(Qt.AlignmentFlag.AlignCenter)
        self.set_tone(tone)

    def set_tone(self, tone: str) -> None:
        if tone not in BADGE_TONES:
            tone = "neutral"
        self._tone = tone
        self.setProperty("tone", tone)
        _repolish(self)

    @property
    def tone(self) -> str:
        return self._tone


class PwbInlineStatus(QWidget):
    """「图标 + 一句话」行内状态（含降级/加载/错误语义，D6 渲染端）。

    tone ∈ {neutral, primary, success, warning, process, error}。
    图标可用 ``icon_name`` 显式覆盖；文字走 QSS 按 tone 着色。
    """

    def __init__(
        self,
        text: str = "",
        *,
        tone: str = "neutral",
        icon_name: str | None = None,
        parent: QWidget | None = None,
    ) -> None:
        super().__init__(parent)
        self._tone = tone
        layout = QHBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        layout.setSpacing(6)
        self._icon_label = QLabel(self)
        self._icon_label.setFixedSize(16, 16)
        self._icon_label.setVisible(False)
        self._text_label = QLabel(text, self)
        self._text_label.setObjectName("PwbInlineStatusText")
        self._text_label.setWordWrap(True)
        layout.addWidget(self._icon_label)
        layout.addWidget(self._text_label, 1)
        layout.addStretch(0)
        self._explicit_icon: str | None = icon_name
        self._apply(text, tone, icon_name)

    def set_status(self, text: str, tone: str | None = None, icon_name: str | None = None) -> None:
        if tone is None:
            tone = self._tone
        self._apply(text, tone, icon_name if icon_name is not None else self._explicit_icon)

    def _apply(self, text: str, tone: str, icon_name: str | None) -> None:
        self._tone = tone
        self._text_label.setText(text)
        self._text_label.setProperty("tone", tone)
        _repolish(self._text_label)
        palette = tokens.palette_for(theme_manager.current_theme.value)
        resolved = icon_name if icon_name is not None else _TONE_ICON.get(tone, ("", ""))[0]
        if resolved:
            color_token = _TONE_ICON.get(tone, ("", "TEXT_SECONDARY"))[1]
            icon = workstation_icon(resolved, str(palette.get(color_token, "")))
            self._icon_label.setPixmap(icon.pixmap(16, 16))
            self._icon_label.setVisible(True)
        else:
            self._icon_label.setVisible(False)

    def clear(self) -> None:
        self._apply("", "neutral", self._explicit_icon)

File: libs/ui_widgets/src/badges.cpp
#include "pwb/ui_widgets/badges.hpp"

#include "pwb/ui_widgets/icon_factory.hpp"
#include "pwb/ui_widgets/ui_context.hpp"

#include <QHBoxLayout>

#include <optional>

namespace pwb::ui_widgets {

const char* const kBadgeTones[] = {"neutral", "primary", "success",
                                   "warning", "error"};

BadgeToneIcon badge_tone_icon(const QString& tone) {
    static const QHash<QString, BadgeToneIcon> table = {
        {"neutral", {"", "TEXT_SECONDARY"}},
        {"primary", {"", "PRIMARY"}},
        {"success", {"circle-check.svg", "SUCCESS"}},
        {"warning", {"alert-triangle.svg", "WARNING"}},
        {"process", {"refresh-cw.svg", "ACCENT"}},
        {"error", {"alert-triangle.svg", "ERROR"}},
    };
    return table.value(tone, {"", "TEXT_SECONDARY"});
}

namespace {

bool is_badge_tone(const QString& tone) {
    for (const char* t : kBadgeTones) {
        if (tone == QLatin1String(t)) return true;
    }
    return false;
}

// Python None-vs-"" semantics: null QString = None (use tone default /
// explicit icon), non-null "" = explicitly no icon.
std::optional<QString> none_if_null(const QString& s) {
    return s.isNull() ? std::nullopt : std::optional<QString>(s);
}

}  // namespace

PwbBadge::PwbBadge(const QString& text, const QString& tone, QWidget* parent)
    : QLabel(text, parent) {
    setObjectName(QStringLiteral("PwbBadge"));
    setAlignment(Qt::AlignCenter);
    set_tone(tone);
}

void PwbBadge::set_tone(const QString& tone) {
    tone_ = is_badge_tone(tone) ? tone : QStringLiteral("neutral");
    setProperty("tone", tone_);
    repolish(this);
}

PwbInlineStatus::PwbInlineStatus(const QString& text, const QString& tone,
                                 const QString& icon_name, QWidget* parent)
    : QWidget(parent), tone_(tone), explicit_icon_(icon_name) {
    auto* layout = new QHBoxLayout(this);
    layout->setContentsMargins(0, 0, 0, 0);
    layout->setSpacing(6);
    icon_label_ = new QLabel(this);
    icon_label_->setFixedSize(16, 16);
    icon_label_->setVisible(false);
    text_label_ = new QLabel(text, this);
    text_label_->setObjectName(QStringLiteral("PwbInlineStatusText"));
    text_label_->setWordWrap(true);
    layout->addWidget(icon_label_);
    layout->addWidget(text_label_, 1);
    layout->addStretch(0);
    apply(text, tone, none_if_null(icon_name));
}

void PwbInlineStatus::set_status(const QString& text, const QString& tone,
                                 const QString& icon_name) {
    const QString effective_tone = tone.isEmpty() ? tone_ : tone;
    apply(text, effective_tone,
          icon_name.isNull() ? none_if_null(explicit_icon_)
                             : std::optional<QString>(icon_name));
}

void PwbInlineStatus::apply(const QString& text, const QString& tone,
                            const QString& icon_name) {
    apply(text, tone, none_if_null(icon_name));
}

void PwbInlineStatus::apply(const QString& text, const QString& tone,
                            const std::optional<QString>& icon_name) {
    tone_ = tone;
    text_label_->setText(text);
    text_label_->setProperty("tone", tone);
    repolish(text_label_);
    const BadgeToneIcon tone_icon = badge_tone_icon(tone);
    const QString resolved =
        icon_name.has_value() ? *icon_name : tone_icon.icon_name;
    if (!resolved.isEmpty()) {
        const QString color =
            palette_token(tone_icon.color_token.toUtf8().constData());
        const QIcon icon = workstation_icon(resolved, color);
        icon_label_->setPixmap(icon.pixmap(16, 16));
        icon_label_->setVisible(true);
    } else {
        icon_label_->setVisible(false);
    }
}

void PwbInlineStatus::clear() {
    apply("", "neutral", none_if_null(explicit_icon_));
}

}  // namespace pwb::ui_widgets
